File: C++/Linked List/34_first_non_repeating_char.cpp
/*
Idea:
using 2 vectors to store frequency of each character & it's last occurrence
While iterating the string-checking if element has occurred before or not & if yes, finding the second unique element occured
& updating the answer string.
*/

//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
    string FirstNonRepeating(string A)
    {
        //storing the size of the string.
        int n = A.size();

        //creating a frequency array for all the characters in the string.
        vector<int> f(26, 0);

        //creating an array to store the last occurrence of each character.
        vector<int> last(26, -1);

        //iterating over the string to update the last occurrence of each character.
        for (int i = 0; i < A.size(); i++)
        {
            if (last[A[i] - 'a'] == -1)
                last[A[i] - 'a'] = i;
        }

        //string to store the answer.
        string ans = "";

        //iterating over the string again to update the frequency of each character and find the first non-repeating character.
        for (int i = 0; i < A.size(); i++)
        {
            f[A[i] - 'a']++;
            //character to store the first non-repeating character.
            char ch = '#';

            //variable to store the last occurrence index of the first non-repeating character.
            int x = A.size() + 1;

            //iterating over the frequency array to find the first non-repeating character.
            for (int j = 0; j < 26; j++)
                if (f[j] == 1 and x > last[j])
                {
                    ch = char(j + 'a');
                    x = last[j];
                }


            //appending the first non-repeating character to the answer string.
            ans += ch;
        }

        //returning the answer string.
        return ans;
    }

};

//{ Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        string A;
        cin >> A;
        Solution obj;
        string ans = obj.FirstNonRepeating(A);
        cout << ans << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Python/Linked List/34_first_non_repeating_char.py
"""
Idea:
using 2 vectors to store frequency of each character & it's last occurrence
While iterating the string-checking if element has occurred before or not & if yes, finding the second unique element occured
& updating the answer string.
"""


# User function Template for python3

class Solution:
    def FirstNonRepeating(self, A):

        # Code here
        # storing the size of the string.
        n = len(A)

        # creating a frequency array for all the characters in the string.
        f = [0] * 26

        # creating an array to store the last occurrence of each character.
        last = [-1] * 26

        # iterating over the string to update the last occurrence of each character.
        for i in range(n):
            if last[ord(A[i]) - ord('a')] == -1:
                last[ord(A[i]) - ord('a')] = i

        # string to store the answer.
        ans = ""

        # iterating over the string again to update the frequency of each character
        # and find the first non-repeating character.
        for i in range(n):
            f[ord(A[i]) - ord('a')] += 1

            # character to store the first non-repeating character.
            ch = '#'

            # variable to store the last occurrence index of the first non-repeating character.
            x = n + 1

            # iterating over the frequency array to find the first non-repeating character.
            for j in range(26):
                if f[j] == 1 and x > last[j]:
                    ch = chr(j + ord('a'))
                    x = last[j]

            # appending the first non-repeating character to the answer string.
            ans += ch

        # returning the answer string.
        return ans


# {
# Driver Code Starts

# Initial Template for Python 3

if __name__ == '__main__':
    T = int(input())
    for i in range(T):
        A = input()
        ob = Solution()
        ans = ob.FirstNonRepeating(A)
        print(ans)

# } Driver Code Ends
